Tutorial: Astro analytics (#7212)
PostHog/posthog.com: contents/tutorials/astro-analytics.md

---
title: How to set up Astro analytics, feature flags, and more
date: 2023-11-24
author: ["ian-vanagas"]
tags: ["configuration", "feature flags", "events"]
---

> Astro is a frontend JavaScript framework focused on performance and simplifying the creation of content-based sites. It has seen a rapid increase in interest and usage since its release in 2022.

Reviewer: ```suggestion
[Astro](https://astro.build/) is a frontend JavaScript framework focused on performance and simplifying the creation of content-based sites. It has seen a rapid increase in interest and usage since its release in 2022.
```

> PostHog provides the tools you need to create the best possible Astro app. In this tutorial, we show you how to set them up. We create a basic Astro blog app, add PostHog on both the client and server, capture custom events, and set up feature flags.

Reviewer: ```suggestion
PostHog provides the tools you need to create the best possible Astro app. In this tutorial, we show you how to set them up. We create a basic Astro blog app, add PostHog on both the client and server, [capture custom events](/docs/product-analytics/capture-events), and set up [feature flags](/docs/feature-flags).
```

## Creating your Astro app

First make sure you install a [Node version](https://nodejs.org/en/download) greater than 18. After that, you can run the command below to create your app. Name your app (we choose `astro-tutorial`), start your new project `Empty`, install dependencies, choose `No` for TypeScript, and `No` for git repository.

> 

Reviewer: For the context of this tutorial, I think it keeps things simpler, and theres not too large of a difference between JS and TS

> 

Author: Yeah, basically my thought is:
1. Most other tutorials are in JS, it's just simpler to write and cleaner to show
2. TS people can figure out JS, but not all JS people know TS

		<title>Astro</title>
	</head>
	<body>
		<a href="/">Home</a>
    <a href="/posts/">Posts</a>
    <slot /> 
	</body>
</html>
```

Next, in the `src/pages` folder, we create a `posts` folder and a basic `index.astro` file with a link to a post we will create.

```js
---
// src/pages/posts/index.astro
import Layout from '../../layouts/Layout.astro';
---
<Layout>
	<h1>Here are all the posts</h1>
	<a href="/posts/first-post">First Post</a>
</Layout>
```

In the `posts` folder, we also create a `first-post.md` file where we write a bit of Markdown as an example. You can customize this or add more posts and links here if you’d like.

```md
---
layout: ../../layouts/Layout.astro
---
<!-- src/pages/posts/first-post.md -->
# First post

PostHog is awesome
```

We also clean up the base `index.astro` file in the pages folder to use the layout we created.

```js
---
// src/pages/index.astro
import Layout from '../layouts/Layout.astro';
---
<Layout>
	<h1>Home</h1>
</Layout>
```

Finally, we can run our app with `npm run dev` to see our full Astro app running locally.

![Astro app video](../images/tutorials/astro-analytics/astro.mp4)

## Adding PostHog on the client side

> With our app set up, we can add PostHog to it. To start, create a new `components` folder in the `src` folder. In this folder, create a `posthog.astro` file. In this file, add your Javascript Web snippet which you can find in [your project settings](https://app.posthog.com/settings/project#snippet). 

Reviewer: ```suggestion
With our app set up, the next step is to add PostHog to it. To start, create a new `components` folder in the `src` folder. In this folder, create a `posthog.astro` file. In this file, add your Javascript Web snippet which you can find in [your project settings](https://app.posthog.com/settings/project#snippet). 
```

  posthog.init(
    '<ph_project_api_key>',
    {
      api_host:'<ph_instance_address>'
    }
  )
</script>
```

After doing this, go back to your `Layout.astro` file, import PostHog, and then add it to the header section.

```js
---
// src/layouts/Layout.astro
import PostHog from '../components/posthog.astro'
---
<html lang="en">
	<head>
		<meta charset="utf-8" />
		<link rel="icon" type="image/svg+xml" href="/favicon.svg" />
		<meta name="viewport" content="width=device-width" />
		<meta name="generator" content={Astro.generator} />
		<title>Astro</title>
    <PostHog />
	</head>
	<body>
		<a href="/">Home</a>
    <a href="/posts/">Posts</a>
    <slot /> 
	</body>
</html>
```

> When you go back to your app and reload, PostHog now autocaptures pageviews, button clicks, session replays (if [you have them enabled](https://app.posthog.com/settings/project-replay)), and more.

Reviewer: ```suggestion
When you go back to your app and reload, PostHog now autocaptures pageviews, button clicks, session replays (if you [enable them](https://app.posthog.com/settings/project-replay)), and more.
```

![Event autocaptured](../images/tutorials/astro-analytics/events.png)

## Capturing custom events

Beyond autocapture, you can use PostHog to capture any events you want. To showcase this, we start by adding a button to our homepage component.

```js
---
// src/pages/index.astro
import Layout from '../layouts/Layout.astro';
---
<Layout>
	<h1>Home</h1>
	<button class="main">Great site!</button>
</Layout>
```

We then add a script to select this button, add a click event listener, and then capture a custom event.

```js
---
// src/pages/index.astro
import Layout from '../layouts/Layout.astro';
---
<Layout>
	<h1>Home</h1>
	<button class="main">Great site!</button>
	<script>
		const button = document.querySelector('.main');
		button.addEventListener('click', () => {
			window.posthog.capture('praise_received')
		});
	</script>
</Layout>
```

When you go back to your app and click the button, you then see a `praise_received` event in PostHog.

![Custom event capture](../images/tutorials/astro-analytics/praise.png)

## Setting up a feature flag

> To use a feature flag in your app, first, you must create it in PostHog. Go to the [feature flags tab](https://app.posthog.com/feature_flags), click "New feature flag," add a key (we choose `new-button`), set the release condition to 100% of users, and click "Save."

Reviewer: ```suggestion
To use a feature flag in your app, first, you must create it in PostHog. Go to the [feature flags tab](https://app.posthog.com/feature_flags), click "New feature flag," add a key (we chose `new-button`), set the release condition to 100% of users, and click "Save."
```

```js
// src/posthog.js
import { PostHog } from 'posthog-node';

let posthogClient = null;

export default function PostHogClient() {
  if (!posthogClient) {
    posthogClient = new PostHog('<ph_project_api_key>', {
      host: '<ph_instance_address>',
    });
  }
  return posthogClient;
}
```

To show how we can use this, we import it into our home `pages/index.astro` file and then capture an event with it.

```js
---
// src/pages/index.astro
import Layout from '../layouts/Layout.astro';
import PostHogClient from '../posthog.js';

const phClient = PostHogClient();
phClient.capture(
	{
		event: 'server_side_event', 
		distinctId: 'ian@posthog.com'
	}
);
---
<!-- The rest of the component -->
```

Now when we reload the page, PostHog captures a `server_side_event`.

## Identifying users to connect server and client

One issue with our implementation is the hard-coded user distinct ID we are using to capture events. For it to use a real user’s ID, we can either:

1. Use the distinct ID set by the client side PostHog in the request cookie.

> 2. Create a new distinct ID on the server and use `identify()` to connect it in the client.

Reviewer: Worth checking if the hyperlink renders correctly. I cant remember whats the correct order for using them in codeblocks
```suggestion
2. Create a new distinct ID on the server and use [`identify()`](/docs/product-analytics/identify) to connect it in the client.
```
